refactor(ivy): move property instruction to a separate location

benlesh opened the pull request:
Simply moves the `property` instruction to its own location, to help with maintenance while we work on migrating to new update instructions.

- Also moves shared functionality we don't want to be exported publicly to the existing `shared.ts` file.

benlesh commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -1338,7 +1268,7 @@ function matchingSchemas(hostView: LView, tagName: string | null): boolean {
  * This enables features like DebugElement.properties.
  */
-function savePropertyDebugData(
+export function savePropertyDebugData(
```
Shoot.. these need to move to `shared` also... VS Code got me :)

kara commented on packages/core/src/render3/instructions/instructions.ts:
````diff
@@ -1093,11 +1097,30 @@ export function elementEnd(): void {
+  * }
+  * ```
+  * @param index the index of the item to act on with the following instructions
+  */
````
Nit: Can we also change the commit message? It sounds like we are only moving the property instruction, but I think the more important part of the PR is that we are adding this select functionality and related tests.

benlesh committed:
refactor(ivy): move property instruction

- moves the property instruction to its own file
- moves shared functions that should not be public to the existing `shared.ts` file.
- adds the export of `property.ts` to `all.ts`

benlesh commented:
blocked on #29513

benlesh commented:
[presubmit](http://test/OCL:240876341:BASE:240876376:1553815160503:8008bdf9)

benlesh commented:
Presubmit failed on circular reference last time. I suspect because VS Code auto-added `import { NO_CHANGE } from '..`;` instead of `from '../tokens'`.  Resolved that and running again.

benlesh committed:
fixup! refactor(ivy): move property instruction

mhevery commented on packages/core/src/render3/instructions/property.ts:
```diff
@@ -0,0 +1,48 @@
+import {getSelectedIndex} from '../state';
+
+import {bind} from './instructions';
+import {elementPropertyInternal} from './shared';
```
Why do we have `elementPropertyInternal` in `shared` and not in this file? 

One concern is that WebPack does not know how to create chunk boundaries across files. So if one symbol in `shared` is needed then the whole file is included in the root chunk, even if the remainder of symbols are not needed until some later lazy loading point. For this reason I think we should not have `shared` which would be dumping ground for things.

benlesh replied:
There are two reasons:

1. It's used in both `instructions.ts` and `property.ts`
2. `all.ts` exports `*` from both `instructions` and `property`. Moving it here would cause it to be exported publicly. We could just explicitly export, but the prior art was to export `*`.

benlesh replied:
Actually, I think I'm going to change this to put everything (including existing `elementProperty` instructions) in `property.ts`, and then my followup PR will put the `propertyInterpolate` instructions in `property.ts` as well.

benlesh replied:
Okay, after our conversation yesterday, I've opened a different PR that moves all of the instructions here: #29646

mhevery commented on packages/core/src/render3/instructions/property.ts:
```diff
@@ -0,0 +1,48 @@
+ * This is a temporary type that will be removed when the issue is resolved.
+ * https://github.com/angular/tsickle/issues/1009)
+ */
+export type TsickleIssue1009 = any;
```
This will be needed in other places as well, (not specific to `property.ts` hence this should be in `shared`)

mhevery commented on packages/core/src/render3/instructions/property.ts:
```diff
@@ -0,0 +1,48 @@
+ *
+ * Operates on the element selected by index via the {@link select} instruction.
+ *
+ * If the property name also exists as an input property on one of the element's directives,
```
This is public API and should be written from the Angular's developer point of view on how this instruction should be used and what it is for. This comment seems to be talking about some implementation detail.

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -34,4 +45,208 @@ export function setInputsForProperty(lView: LView, inputs: PropertyAliasValue, v
+/**
+ * Mapping between attributes names that don't correspond to their element property names.
+ */
+const ATTR_TO_PROP: {[name: string]: string} = {
```
I think just about everything should go to `property.ts` per reasons discussed.

benlesh replied:
Yeah, please see the other PR here: #29646  (per our discussion yesterday)

mhevery commented on packages/core/src/render3/instructions/property.ts:
```diff
@@ -0,0 +1,48 @@
+ * @returns This function returns itself so that it may be chained
+ * (e.g. `property('name', ctx.name)('title', ctx.title)`)
+ */
+export function property<T>(
```
rename `property` to `®property`?

benlesh replied:
I've started one PR to move the instructions here: #29646

... I'd like to rename everything in a followup PR to that.

mhevery reviewed (approved):
LGTM, but please address my comments.

benlesh commented:
Closing this in favor of #29646
